{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMlmGkY5VhXAwglLoBxEXRt",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mikemudgett/APHID/blob/main/APHID_online.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%capture\n",
        "# @title Install and import libraries\n",
        "\n",
        "# @markdown Press the play button on the left to install the needed libraries.\n",
        "# @markdown Nothing is downloaded to your own computer.\n",
        "\n",
        "!apt install ncbi-blast+\n",
        "\n",
        "!pip install primer3-py\n",
        "!pip install Bio\n",
        "!pip install python-docx\n",
        "\n",
        "import sys, os\n",
        "import primer3\n",
        "import pandas as pd\n",
        "from Bio.Blast.Applications import NcbiblastnCommandline\n",
        "from Bio.Blast.Applications import NcbimakeblastdbCommandline\n",
        "import numpy as np\n",
        "import regex as re\n",
        "from Bio.Seq import Seq\n",
        "from Bio import SeqIO\n",
        "from io import StringIO\n",
        "import regex as re\n",
        "from docx import Document\n",
        "from docx.shared import RGBColor"
      ],
      "metadata": {
        "id": "O5sb_g32KvdO",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "aWp1sld5ihh_"
      },
      "outputs": [],
      "source": [
        "# @title Load utility functions\n",
        "\n",
        "# @markdown Load utility functions\n",
        "\n",
        "class Gene:\n",
        "\tdef __init__(self):\n",
        "\t\tself.coding_exons = []\n",
        "\t\tself.five_prime_UTR = []\n",
        "\t\tself.three_prime_UTR = []\n",
        "\n",
        "\tdef set_polarity(self, strand):\n",
        "\t\tif strand == '+':\n",
        "\t\t\tself.polarity = 0\n",
        "\t\telse:\n",
        "\t\t\tself.polarity = -1\n",
        "\n",
        "\tdef set_coords(self, start, end):\n",
        "\t\tself.mRNA_start = start\n",
        "\t\tself.mRNA_end = end\n",
        "\n",
        "\tdef add_feature(self, line):\n",
        "\t\tfields = line.split('\\t')\n",
        "\t\tmatch fields[2]:\n",
        "\t\t\tcase 'mRNA':\n",
        "\t\t\t\tself.set_polarity(fields[6])\n",
        "\t\t\t\tself.set_coords(int(fields[3]), int(fields[4]))\n",
        "\t\t\tcase 'exon':\n",
        "\t\t\t\tpass\n",
        "\t\t\tcase 'CDS':\n",
        "\t\t\t\t#Start, End, Length\n",
        "\t\t\t\t#Assumes the exons are all in order\n",
        "\t\t\t\tself.coding_exons.append((int(fields[3]), int(fields[4]), int(fields[4]) - int(fields[3])+1))\n",
        "\t\t\tcase 'five_prime_UTR':\n",
        "\t\t\t\tself.five_prime_UTR.append((int(fields[3]), int(fields[4])))\n",
        "\t\t\tcase 'three_prime_UTR':\n",
        "\t\t\t\tself.three_prime_UTR.append((int(fields[3]), int(fields[4])))\n",
        "\t\t\tcase _:\n",
        "\t\t\t\tpass\n",
        "\n",
        "#DOES NOT give the reverse complement if you have '-' direction.\n",
        "def grow_primer_to_tm(seq, start_pos, direction, min_Tm=59, min_length=15):\n",
        "\tmelt_temp = 0\n",
        "\tincrement = min_length\n",
        "\tstart = start_pos\n",
        "\twhile melt_temp < min_Tm:\n",
        "\t\tif direction == '+':\n",
        "\t\t\tprimer = seq[start_pos:start+increment+1]\n",
        "\t\t\tmelt_temp = primer3.calc_tm(str(primer))\n",
        "\t\telse:\n",
        "\t\t\tprimer = seq[start+1-increment:start+1]\n",
        "\t\t\tmelt_temp = primer3.calc_tm(str(primer))\n",
        "\n",
        "\t\tincrement += 1\n",
        "\t#This returns a \"Seq\" not a string. A little sloppy.\n",
        "\treturn primer\n",
        "\n",
        "\n",
        "#BLAST each candidate we have at this point against the whole genome database.\n",
        "#Maybe just select the candidate with the least number of significant hits?\n",
        "#Make the final decision based on the longest homology arm.\n",
        "def get_blast_hits(seq):\n",
        "\n",
        "#for candidate in candidate_primer_list.copy():\n",
        "\ttemp_fa = open('candidate.fa', 'w')\n",
        "\ttemp_fa.write('>candidate' + '\\n' + seq)\n",
        "\ttemp_fa.close()\n",
        "\tblast_cline = NcbiblastnCommandline(query='candidate.fa',db='athal', evalue=10, task='blastn-short', outfmt='10 nident qstart qend')\n",
        "\tblast_result = pd.read_csv(StringIO(blast_cline()[0]))\n",
        "\t#print(len(blast_result))\n",
        "\t#return len(blast_result)\n",
        "\t#return blast_result.sum()[1]\n",
        "\treturn blast_result\n",
        "\n",
        "\n",
        "#Figure out the strandedness yet. I don't know if there is an issue here.\n",
        "#Faster now.\n",
        "#Everything might kind of be off by one. Might not really matter?\n",
        "def choose_primer(template_seq, lower_bound, upper_bound, strand, opposite_primer='', min_Tm=55, max_Tm=62, min_length=20, max_length=27):\n",
        "\n",
        "\tprimer_candidate_list = pd.DataFrame(columns = ['sequence', 'local_position', 'length', 'tm', 'hairpin_tm', 'homodimer_tm', 'heterodimer_tm', 'thermo_tm_sum', 'blast'])\n",
        "\n",
        "\tmin_pos = lower_bound\n",
        "\tmax_pos = upper_bound\n",
        "\n",
        "\tfor length in range(min_length, max_length+1):\n",
        "\t\tif strand == '+':\n",
        "\t\t\tstart = min_pos\n",
        "\t\t\tstop = max_pos\n",
        "\t\telif strand == '-':\n",
        "\t\t\tstart = min_pos - length\n",
        "\t\t\tstop = max_pos - length\n",
        "\t\t#Get BLAST hits for the entire range of possible primers of this length.\n",
        "\t\tblast_result = get_blast_hits(str(template_seq[start:stop + length]))\n",
        "\t\tfor x in range(start, stop):\n",
        "\t\t\tcand_seq = str(template_seq[x:x+length])\n",
        "\t\t\tif strand == '-':\n",
        "\t\t\t\tcand_seq = str(Seq(cand_seq).reverse_complement())\n",
        "\t\t\tcand_tm = primer3.calc_tm(cand_seq)\n",
        "\t\t\tif (cand_tm > min_Tm) and (cand_tm < max_Tm):\n",
        "\t\t\t\tcand_hairpin_tm = primer3.calc_hairpin(cand_seq, dna_conc=200).tm\n",
        "\t\t\t\tcand_homodimer_tm = primer3.calc_homodimer(cand_seq, dna_conc=200).tm\n",
        "\t\t\t\tif opposite_primer == '':\n",
        "\t\t\t\t\tcand_heterodimer_tm = 0\n",
        "\t\t\t\telse:\n",
        "\t\t\t\t\tcand_heterodimer_tm = primer3.calc_heterodimer(cand_seq, opposite_primer, dna_conc=200).tm\n",
        "\t\t\t\t#Don't need to bother with meeting a threshold if we aren't blasting every single one of them.\n",
        "\t\t\t\t#if (cand_hairpin_tm < thermo_thresh) and (cand_homodimer_tm < thermo_thresh) and (cand_heterodimer_tm < thermo_thresh):\n",
        "\t\t\t\t\t#cand_blast = get_blast_hits(cand_seq)\n",
        "\n",
        "\n",
        "\t\t\t\t#Now we just have to figure out how to get the number of blast hits that match a given primer\n",
        "\t\t\t\tcand_blast = 0\n",
        "\t\t\t\tfor entry in blast_result.index:\n",
        "\t\t\t\t\t#print(str(x) + '  ' + str(blast_result.iloc[entry,1]))\n",
        "\n",
        "\t\t\t\t\tif ((x+1-start)<=blast_result.iloc[entry,1]) and ((x+length-start)>=blast_result.iloc[entry,2]):\n",
        "\t\t\t\t\t\tcand_blast += 1\n",
        "\n",
        "\n",
        "\t\t\t\tprimer_candidate_list = pd.concat([primer_candidate_list, pd.Series({'sequence': cand_seq,\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'local_position': int(x),\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'length': int(length),\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'tm': cand_tm,\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'hairpin_tm': cand_hairpin_tm,\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'homodimer_tm': cand_homodimer_tm,\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'heterodimer_tm': cand_heterodimer_tm,\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'thermo_tm_sum': cand_hairpin_tm + cand_homodimer_tm + cand_heterodimer_tm,\n",
        "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t 'blast': cand_blast}).to_frame().T], ignore_index = True,)\n",
        "\n",
        "\tsort_list = ['blast', 'thermo_tm_sum', 'length', 'local_position']\n",
        "\tif strand == '+':\n",
        "\t\tsort_list_order = [True, True, True, True]\n",
        "\telif strand == '-':\n",
        "\t\tsort_list_order = [True, True, True, False]\n",
        "\n",
        "\tprimer_candidate_list = primer_candidate_list.sort_values(sort_list, ascending=sort_list_order)\n",
        "\n",
        "\treturn primer_candidate_list.iloc[0]\n",
        "\t#this is a pandas series with all the info generated about the chosen primer\n",
        "\n",
        "\n",
        "def pull_guides(chr_seq, pam_seq, guide_length):\n",
        "\tIUPAC_dict = {'A':'A', 'C':'C', 'G':'G', 'T':'T', 'U':'T',\n",
        "\t\t\t\t  'R':'[AG]', 'Y':'[CT]', 'S':'[GC]', 'W':'[AT]',\n",
        "\t\t\t\t  'K':'[GT]', 'M':'[AC]', 'B':'[CGT]', 'D':'[AGT]',\n",
        "\t\t\t\t  'H':'[ACT]', 'V':'[ACG]', 'N':'[ACGT]', 'X':'[ACGT]'}\n",
        "\n",
        "\tguide_list = []\n",
        "\n",
        "\tregex_token = '.{' + str(guide_length) + '}'\n",
        "\tfor x in pam_seq:\n",
        "\t\tregex_token += IUPAC_dict[x]\n",
        "\n",
        "\n",
        "\trc_pam_seq = Seq(pam_seq).reverse_complement()\n",
        "\trc_regex_token = ''\n",
        "\tfor x in rc_pam_seq:\n",
        "\t\trc_regex_token += IUPAC_dict[x]\n",
        "\trc_regex_token += '.{' + str(guide_length) + '}'\n",
        "\n",
        "\tsites = re.findall(regex_token, chr_seq, overlapped=True)\n",
        "\tfor guide in sites:\n",
        "\t\tguide_list.append(guide[0:guide_length])\n",
        "\n",
        "\trc_sites = re.findall(rc_regex_token, chr_seq, overlapped=True)\n",
        "\tfor rc_guide in rc_sites:\n",
        "\t\tguide_list.append(str(Seq(rc_guide).reverse_complement())[0:guide_length])\n",
        "\n",
        "\treturn guide_list"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Run prep functions\n",
        "\n",
        "# @markdown Initial functions to build guide library and BLAST database.<br>\n",
        "# @markdown Nothing is downloaded to your own computer.<br>\n",
        "# @markdown Running may take a few minutes.\n",
        "\n",
        "# @markdown\n",
        "\n",
        "\n",
        "#This script prepares the necessary files to run APHID. It comprises of three steps:\n",
        "#1. Download the genome files if they are not already in the current directory.\n",
        "#2. Pull a list of all the guides in the genome, specific to your PAM sequences and guide length.\n",
        "#   This will make it easier to search for possible off-targets.\n",
        "#3. Create a BLAST database of the Arabidopsis genome. This will make finding specific primers easier.\n",
        "\n",
        "\n",
        "pam_seq = 'NGG' #@param {type:\"string\"}\n",
        "guide_length = 20 #@param {type:\"integer\"}\n",
        "\n",
        "#PART 1: Genome Download\n",
        "\n",
        "#Download individual chromosome files if they are not already in the directory.\n",
        "#We are using Arabidopsis thaliana release 56 from ensembl plants. This could be updated with new releases as long as the naming conventions stay the same.\n",
        "print('Downloading genome and annotation files...')\n",
        "for num in ['1', '2', '3', '4', '5', 'Mt', 'Pt']:\n",
        "\tfn = 'Arabidopsis_thaliana.TAIR10.dna.chromosome.' + num + '.fa'\n",
        "\tif not os.path.exists(fn):\n",
        "\t\tos.system('wget -nc https://ftp.ebi.ac.uk/ensemblgenomes/pub/release-56/plants/fasta/arabidopsis_thaliana/dna/' + fn + '.gz')\n",
        "\t\tos.system('gunzip ' + fn)\n",
        "\tfn = 'Arabidopsis_thaliana.TAIR10.56.chromosome.' + num + '.gff3'\n",
        "\tif not os.path.exists(fn):\n",
        "\t\tos.system('wget -nc https://ftp.ebi.ac.uk/ensemblgenomes/pub/release-56/plants/gff3/arabidopsis_thaliana/' + fn + '.gz')\n",
        "\t\tos.system('gunzip ' + fn)\n",
        "print('Genome files are downloaded.')\n",
        "print('')\n",
        "\n",
        "#PART 2: Guide Search\n",
        "\n",
        "#this is pretty slow, takes ~7 minutes for NGG and 20bp guides\n",
        "print('Creating list of all nuclear guide sites with PAM ' + pam_seq + ' of length ' + str(guide_length) + '...')\n",
        "\n",
        "out_fn = 'all_guides_' + pam_seq + '_' + str(guide_length) + '.txt'\n",
        "\n",
        "if os.path.exists(out_fn):\n",
        "\tprint('Guide list already exists in directory. Moving on to creating BLAST database.')\n",
        "else:\n",
        "\tout_file = open(out_fn, 'w')\n",
        "\tprint('Scanning genome for guides.')\n",
        "\tfor chr_num in (1,2,3,4,5):\n",
        "\t\twith open('Arabidopsis_thaliana.TAIR10.dna.chromosome.' + str(chr_num) + '.fa') as fa:\n",
        "\t\t\tfor record in SeqIO.parse(fa, \"fasta\"):\n",
        "\t\t\t\tchr_seq = str(record.seq)\n",
        "\t\t\t\tchr_guides = pull_guides(chr_seq, pam_seq, guide_length)\n",
        "\t\t\t\tfor guide in chr_guides:\n",
        "\t\t\t\t\tout_file.write(str(guide) + '\\n')\n",
        "\t\tprint('Completed scanning chromosome ' + str(chr_num) + '.')\n",
        "\n",
        "\tout_file.close()\n",
        "\tprint('Guide list complete.')\n",
        "print('')\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "#PART 3: BLAST Database\n",
        "print('Building BLAST database...')\n",
        "\n",
        "\n",
        "if os.path.exists('Arabidopsis_thaliana.TAIR10.dna.full.fa'):\n",
        "\tprint('Concatenated genome already in directory.')\n",
        "else:\n",
        "\tprint('Concatenating genome.')\n",
        "\tos.system('cat Arabidopsis_thaliana.TAIR10.dna.chromosome.*.fa > Arabidopsis_thaliana.TAIR10.dna.full.fa')\n",
        "\n",
        "if os.path.exists('athal.nhr'):\n",
        "\tprint('BLAST database already exists in directory.')\n",
        "else:\n",
        "\tprint('Generating database.')\n",
        "\tmakedb_cline = NcbimakeblastdbCommandline(dbtype='nucl', input_file='Arabidopsis_thaliana.TAIR10.dna.full.fa', out='athal')\n",
        "\tmakedb_cline()\n",
        "\n",
        "print('')\n",
        "print('Setup complete. You may move on to using APHID.')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "fB6V9KFMMiMT",
        "outputId": "45b45637-e929-4283-dc66-ae367413995d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading genome and annotation files...\n",
            "Genome files are downloaded.\n",
            "\n",
            "Creating list of all nuclear guide sites with PAM NGG of length 20...\n",
            "Scanning genome for guides.\n",
            "Completed scanning chromosome 1.\n",
            "Completed scanning chromosome 2.\n",
            "Completed scanning chromosome 3.\n",
            "Completed scanning chromosome 4.\n",
            "Completed scanning chromosome 5.\n",
            "Guide list complete.\n",
            "\n",
            "Building BLAST database...\n",
            "Concatenating genome.\n",
            "Generating database.\n",
            "\n",
            "Setup complete. You may move on to using APHID.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Run APHID\n",
        "\n",
        "# @markdown Run the APHID program to design your insertion and construct.<br>\n",
        "# @markdown The only required file you must provide is the cargo sequence, which should be in FASTA format (.fa) and included in this directory.<br>\n",
        "\n",
        "# @markdown After updating the four required parameters below, you can run this step.<br>\n",
        "# @markdown Once it has completed running, the output files should download to your computer.\n",
        "transcript_name = 'AT1G73590.1' #@param {type:\"string\"}\n",
        "insertion_location = 452 #@param {type:\"integer\"}\n",
        "cargo_fn = 'GFP.fa' #@param {type:\"string\"}\n",
        "cargo_name = cargo_fn.split('.')[0]\n",
        "project_name = 'PIN1-GFP' #@param {type:\"string\"}\n",
        "\n",
        "try:\n",
        "\tcargo_file = open(cargo_fn, 'r')\n",
        "except:\n",
        "\tprint('Error! Could not find cargo file.')\n",
        "\tsys.exit()\n",
        "\n",
        "#VARIABLES\n",
        "\n",
        "# @markdown **All the other variables are set to the recommended defaults**,\n",
        "# @markdown but can be changed according to your needs.\n",
        "\n",
        "local_seq_range = 1500 #How much sequence to look at on either side of the insertion site. Do not change unless you are using very large homology arms or want more sequence in the final output.\n",
        "\n",
        "# @markdown The following parameters relate to your nuclease. Default is SpCas9.\n",
        "\n",
        "#length of guides in base pairs\n",
        "guide_length = 20 #@param {type:\"integer\"}\n",
        "\n",
        "#PAM recognition sequence \"NGG\" for SpCas9\n",
        "pam_seq = 'NGG' #@param {type:\"string\"}\n",
        "\n",
        "#Predicted nuclease cut site relative to the PAM. 3bp for SpCas9 (number of basepairs upstream of the PAM)\n",
        "cut_site_shift = 3 #@param {type:\"integer\"}\n",
        "\n",
        "# @markdown The following parameters relate to the size of your desired homology arms (in basepairs)\n",
        "\n",
        "\n",
        "#min and max base pair lengths for homology arms\n",
        "homology_arm_min = 300 #@param {type:\"integer\"}\n",
        "homology_arm_max = 600 #@param {type:\"integer\"}\n",
        "\n",
        "# @markdown The following parameter is the minimum distance (in basepairs) your cut site can be from an \"off-limits\" genomic region. It is padding to allow for deletions.\n",
        "\n",
        "\n",
        "#Number of basepairs padding to include between predicted cut site and nearby relevant genomic features. Default is 10\n",
        "\t\t\t\t   #This is because even if your guide is in an intron, if the cut site is too close, you can disrupt the splice site.\n",
        "indel_padding = 10 #@param {type:\"integer\"}\n",
        "\n",
        "# @markdown The following parameters indicate which areas of the genome guide sites can be designed in (and subjected to indels)\n",
        "\n",
        "\n",
        " #These values determine which regions can be cut and subjected to indels. By default, introns and intergenic sequences are the only ones allowed.\n",
        "cut_intergenic = True #@param {type:\"boolean\"}\n",
        "cut_CDS = False #@param {type:\"boolean\"}\n",
        "cut_three_prime_UTR = False #@param {type:\"boolean\"}\n",
        "cut_intron = True #@param {type:\"boolean\"}\n",
        "cut_five_prime_UTR = False #@param {type:\"boolean\"}\n",
        "cut_other_RNA = False #@param {type:\"boolean\"}\n",
        "\n",
        "gibson_assembly_overhang = 20 #The number of base pair overhang used in Gibson Assembly. 20bp works just fine.\n",
        "\n",
        "gt_space_min = 60 #The minimum and maximum base pair distances from the ends of each homology arm for genotyping primers\n",
        "gt_space_max = 200 #If min is too low, you cannot get good coverage of the junciton during sequencing and the primer binding site could be disrupted by indels.\n",
        "\t\t\t\t   #If max is too high, the PCR product length may be too long to get a reliable band.\n",
        "\n",
        "gt1_opposite = '' #Input sequences here if you have specific genotyping primers for your cargo that you want to make sure are compatible with the genomic genotyping primers.\n",
        "gt2_opposite = '' #If these are blank, the cargo cloning primers will be used.\n",
        "\t\t\t\t  #For example, if your cargo is GFP and you have your own GFP primers to use with GT1 and GT2 respectively, enter them here.\n",
        "\t\t\t\t  #Otherwise, we will assume you are genotyping using GT1 + cargo_R and GT2 + cargo_F.\n",
        "\n",
        "#EXTRACT TRANSCRIPT INFORMATION\n",
        "\n",
        "#The chromosome number is always the third digit in a gene name.\n",
        "chr_num = transcript_name[2]\n",
        "\n",
        "print('Transcript ID: ' + transcript_name)\n",
        "print('Chromosome number: ' + str(chr_num))\n",
        "print('Extracting transcript features...')\n",
        "\n",
        "#Look at the specific annotation file for this chromosome.\n",
        "gff_fn = 'Arabidopsis_thaliana.TAIR10.56.chromosome.' + chr_num + '.gff3'\n",
        "gff = open(gff_fn, 'r')\n",
        "\n",
        "#Create a gene object.\n",
        "gene = Gene()\n",
        "for line in gff:\n",
        "\t#Extract information about the transcript from the annotation file.\n",
        "\tif transcript_name in line:\n",
        "\t\tgene.add_feature(line)\n",
        "gff.close()\n",
        "\n",
        "if gene.coding_exons:\n",
        "\tprint('Complete.')\n",
        "\tprint('')\n",
        "else:\n",
        "\tprint('Error! Transcript details not found in GFF file.')\n",
        "\tsys.exit()\n",
        "\n",
        "\n",
        "\n",
        "#DETERMINE INSERTION COORDINATE\n",
        "\n",
        "#Multiply by three to go from amino acid to nucleotide.\n",
        "cds_insertion_location = insertion_location * 3\n",
        "\n",
        "#If -1 was the insertion location we just want to put it at the end of the gene.\n",
        "if cds_insertion_location < 0:\n",
        "\tprint('Desired inseriton location is at the end of the coding sequence.')\n",
        "\t#Find the end of the coding sequence.\n",
        "\t#The end of this line is adding an offset and a shift of 3 to account for the stop codon.\n",
        "\tinsertion_coordinate = gene.coding_exons[-1][gene.polarity + 1] + gene.polarity + ((-2 * gene.polarity - 1) * 3)\n",
        "\n",
        "#If a non-negative number was given, we have to find the exact insertion site.\n",
        "else:\n",
        "\tprint('Desired insertion is after residue number: ' + str(insertion_location) + '.')\n",
        "\texon_counter = 0\n",
        "\t#Might need a better condition now   CHECK THIS??\n",
        "\twhile cds_insertion_location >= 0:\n",
        "\t\tif gene.coding_exons[exon_counter][2] >= cds_insertion_location:\n",
        "\t\t#If the insertion site fits in this exon, find the correct location.\n",
        "\t\t\t#Once again, little offsets are required to make this work for both positive and negative strand genes and make sure everything is not off by one.\n",
        "\t\t\tinsertion_coordinate = gene.coding_exons[exon_counter][-1 * gene.polarity] + ((2 * gene.polarity + 1) * cds_insertion_location) + ((gene.polarity + 1) * -1)\n",
        "\n",
        "\t\t\t#To leave the loop.\n",
        "\t\t\tcds_insertion_location = -1\n",
        "\t\telse:\n",
        "\t\t\t#Subtract the length of this exon and move on to the next one.\n",
        "\t\t\tcds_insertion_location -= gene.coding_exons[exon_counter][2]\n",
        "\t\t\texon_counter += 1\n",
        "print('Insertion location is at coordinate ' + str(insertion_coordinate) + ' on Chromosome ' + str(chr_num) + '.')\n",
        "print('')\n",
        "\n",
        "#To Do: Add something above to catch if the number is too high.\n",
        "\n",
        "\n",
        "#PULL NUCLEOTIDE SEQUENCE\n",
        "\n",
        "#We are only interested in the sequence flanking the insertion site so we can find guides and primers.\n",
        "print('Extracting sequence flanking the insertion site...')\n",
        "range_min = insertion_coordinate - local_seq_range\n",
        "range_max = insertion_coordinate + local_seq_range\n",
        "\n",
        "with open('Arabidopsis_thaliana.TAIR10.dna.chromosome.' + chr_num + '.fa') as fa:\n",
        "\tfor record in SeqIO.parse(fa, \"fasta\"):\n",
        "\t\tlocal_seq = record.seq[range_min-1:range_max-1]\n",
        "\n",
        "#It should already be uppercase, but we do this to be safe.\n",
        "local_seq = local_seq.upper()\n",
        "#Get the reverse complement of the sequence if the gene of interest is on the negative strand.\n",
        "if gene.polarity == -1:\n",
        "\tlocal_seq = str(Seq(local_seq).reverse_complement())\n",
        "\n",
        "\n",
        "#ENCODE FEATURES IN GENOMIC SEQUENCE\n",
        "\n",
        "#We want an array of integers, where each integer corresponds to an individual nucleotide in the sequence.\n",
        "#The value of the integer indicates what type of feature the nucleotide belongs to.\n",
        "\n",
        "#Sequence Code:\n",
        "#1: Intergenic\n",
        "#2: Coding Sequence\n",
        "#3: 3' UTR\n",
        "#4: Intron\n",
        "#5: 5' UTR\n",
        "#6: Other gene type (lncRNA, miRNA, tRNA, etc.)\n",
        "print('Annotating the sequence around the insertion site...')\n",
        "seq_code = np.ones(len(local_seq))\n",
        "local_range = range(range_min, range_max)\n",
        "\n",
        "#I DONT THINK THIS IS NECESSARY, DELETE IT AFTER TESTING IF IT IS NOT\n",
        "# #Make sure everything is not all off by 1 or something. They are off, figure it out\n",
        "# for coords in gene.five_prime_UTR:\n",
        "#     five_prime_UTR_range = range(coords[0], coords[1]+1)\n",
        "#     overlap = range(max(local_range[0], five_prime_UTR_range[0]), min(local_range[-1], five_prime_UTR_range[-1])+1)\n",
        "#     for x in overlap:\n",
        "#         seq_code[x-range_min] = 5\n",
        "\n",
        "# for coords in gene.three_prime_UTR:\n",
        "#     three_prime_UTR_range = range(coords[0], coords[1]+1)\n",
        "#     overlap = range(max(local_range[0], three_prime_UTR_range[0]), min(local_range[-1], three_prime_UTR_range[-1])+1)\n",
        "#     for x in overlap:\n",
        "#         seq_code[x-range_min] = 3\n",
        "\n",
        "# for coords in gene.coding_exons:\n",
        "#     exon_range = range(coords[0], coords[1]+1)\n",
        "#     overlap = range(max(local_range[0], exon_range[0]), min(local_range[-1], exon_range[-1])+1)\n",
        "#     for x in overlap:\n",
        "#         seq_code[x-range_min] = 2\n",
        "\n",
        "# mRNA_range = range(gene.mRNA_start, gene.mRNA_end+1)\n",
        "# overlap = range(max(local_range[0], mRNA_range[0]), min(local_range[-1], mRNA_range[-1])+1)\n",
        "# for x in overlap:\n",
        "#     if seq_code[x-range_min] == 1:\n",
        "#         seq_code[x-range_min] = 4\n",
        "\n",
        "\n",
        "#We want to scan through the GFF annotation file again and pull whatever information there is about the nulceotides in our local sequence\n",
        "with open('Arabidopsis_thaliana.TAIR10.56.chromosome.' + chr_num + '.gff3') as gff:\n",
        "\tfor line in gff:\n",
        "\t\tfields = line.split('\\t')\n",
        "\t\t#We want to ignore the text lines at the beginning of the file\n",
        "\t\tif len(fields) > 2:\n",
        "\t\t\t#It is impossible to prevent all conflicts within the sequence code since there are overlapping transcripts and such, but the\n",
        "\t\t\t#main thing we want to avoid is issues with our specific gene. This can especially happen if there are different transcripts\n",
        "\t\t\t#in our gene which use sequences differentyl (i.e. retained intron). Thus we want to ignore other transcript IDs from our gene.\n",
        "\t\t\tif not ((transcript_name.split('.')[0] in fields[8]) and (transcript_name not in fields[8])):\n",
        "\t\t\t\t#If the feature described in the line overlaps the local sequence, we want to annotate our sequence code\n",
        "\t\t\t\toverlap = range(max(local_range[0], int(fields[3])), min(local_range[-1], int(fields[4]))+1)\n",
        "\t\t\t\t#x will be any coordinates that overlap\n",
        "\t\t\t\tfor x in overlap:\n",
        "\t\t\t\t\t#This is so that we don't overwrite anything accidentally.\n",
        "\t\t\t\t\t#It is a little tricky because CDS corresponds to exons and mRNA corresponds to exons and introns,\n",
        "\t\t\t\t\t#so we want to make sure that introns can be overwritten but exons cannot.\n",
        "\t\t\t\t\tif seq_code[x-range_min] not in (2, 3, 5, 6):\n",
        "\t\t\t\t\t\tmatch fields[2]:\n",
        "\t\t\t\t\t\t\tcase 'CDS':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 2\n",
        "\t\t\t\t\t\t\tcase 'three_prime_UTR':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 3\n",
        "\t\t\t\t\t\t\tcase 'five_prime_UTR':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 5\n",
        "\t\t\t\t\t\t\tcase 'lnc_RNA':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 6\n",
        "\t\t\t\t\t\t\tcase 'miRNA':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 6\n",
        "\t\t\t\t\t\t\tcase 'ncRNA':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 6\n",
        "\t\t\t\t\t\t\tcase 'snRNA':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 6\n",
        "\t\t\t\t\t\t\tcase 'snoRNA':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 6\n",
        "\t\t\t\t\t\t\tcase 'tRNA':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 6\n",
        "\t\t\t\t\t\t\tcase 'mRNA':\n",
        "\t\t\t\t\t\t\t\tseq_code[x-range_min] = 4\n",
        "\t\t\t\t\t\t\tcase _:\n",
        "\t\t\t\t\t\t\t\tpass\n",
        "\n",
        "#If the gene is on the negative strand, we flip the sequence code so that it matches up with the reverse complemented local sequence.\n",
        "#From now on we will not have to use coordinates from the genome files/annotation so everything can be based off our chosen local sequence.\n",
        "if gene.polarity == -1:\n",
        "\tseq_code = np.flip(seq_code)\n",
        "print('Complete.')\n",
        "print('')\n",
        "\n",
        "\n",
        "#FIND ALL ELIGIBLE GUIDE SITES AND CHOOSE BEST GUIDE\n",
        "\n",
        "local_center = int(len(local_seq) / 2) + (gene.polarity * 2)\n",
        "\n",
        "#Find sites in the left arm\n",
        "left_lower = local_center - homology_arm_max\n",
        "left_upper = local_center - homology_arm_min\n",
        "\n",
        "right_lower = local_center + homology_arm_min\n",
        "right_upper = local_center + homology_arm_max\n",
        "\n",
        "left_arm_seq = local_seq[left_lower:left_upper]\n",
        "right_arm_seq = local_seq[right_lower:right_upper]\n",
        "\n",
        "print('Searching for eligible guide sequences within the desired homology arm regions...')\n",
        "left_guide_candidates = pull_guides(str(left_arm_seq), pam_seq, guide_length)\n",
        "right_guide_candidates = pull_guides(str(right_arm_seq), pam_seq, guide_length)\n",
        "\n",
        "guide_candidate_list = pd.DataFrame(columns = ['sequence', 'local_position', 'strand', 'arm', 'GC', 'TTTT', 'safe_to_cut'])\n",
        "\n",
        "print('Annotating eligible guides and calculating specificity...')\n",
        "#Add details about each guide manually.\n",
        "#First we will add in information about the arm, the strandedness, and the start position of each guide.\n",
        "for guide in left_guide_candidates:\n",
        "\t#We want to know the directionality of the guide. Here, + or - is relative to our local sequence, which may be flipped relative to the standardized genome.\n",
        "\t#We also want to know where the guide starts in our sequence so we can find its position when making primers\n",
        "\tif guide in str(left_arm_seq):\n",
        "\t\tguide_strand = '+'\n",
        "\t\tguide_start = str(local_seq).find(guide)\n",
        "\telse:\n",
        "\t\tguide_strand = '-'\n",
        "\t\t#The position for the reverse complement guides takes a little extra work.\n",
        "\t\tguide_start = (local_seq_range * 2) - str(Seq(local_seq).reverse_complement()).find(guide) - guide_length\n",
        "\n",
        "\tguide_candidate_list = pd.concat([guide_candidate_list, pd.Series({'sequence': guide, 'local_position': guide_start, 'strand': guide_strand, 'arm': 'left'}).to_frame().T], ignore_index = True)\n",
        "\n",
        "#Do the same for the right arm guides.\n",
        "for guide in right_guide_candidates:\n",
        "\tif guide in str(right_arm_seq):\n",
        "\t\tguide_strand = '+'\n",
        "\t\tguide_start = str(local_seq).find(guide)\n",
        "\telse:\n",
        "\t\tguide_strand = '-'\n",
        "\t\tguide_start = (local_seq_range * 2) - str(Seq(local_seq).reverse_complement()).find(guide) - guide_length\n",
        "\n",
        "\tguide_candidate_list = pd.concat([guide_candidate_list, pd.Series({'sequence': guide, 'local_position': guide_start, 'strand': guide_strand, 'arm': 'right'}).to_frame().T], ignore_index = True)\n",
        "\n",
        "#Based on the types of sequence which the user has approved can be subjected to indels, make a list of allowed codes.\n",
        "#This is based on the sequence code from the previous section.\n",
        "allowed_codes = []\n",
        "if cut_intergenic:\n",
        "\tallowed_codes.append(1)\n",
        "if cut_CDS:\n",
        "\tallowed_codes.append(2)\n",
        "if cut_three_prime_UTR:\n",
        "\tallowed_codes.append(3)\n",
        "if cut_intron:\n",
        "\tallowed_codes.append(4)\n",
        "if cut_five_prime_UTR:\n",
        "\tallowed_codes.append(5)\n",
        "if cut_other_RNA:\n",
        "\tallowed_codes.append(6)\n",
        "\n",
        "#Add more information for each guide: TTTT presence, GC content, and permission to cut.\n",
        "for entry in guide_candidate_list.index:\n",
        "\n",
        "\tguide_seq = guide_candidate_list.loc[entry].sequence\n",
        "\tguide_strand = guide_candidate_list.loc[entry].strand\n",
        "\tguide_start = guide_candidate_list.loc[entry].local_position\n",
        "\n",
        "\t#Flag guide sequences with TTTT, which is a terminating signal for RNA Pol III.\n",
        "\t#This wouldn't be necessary if you are not transcribing gRNAs with RNA Pol III (i.e. You are not using a U6 promoter).\n",
        "\tif ('TTTT' in guide_seq):\n",
        "\t\tguide_candidate_list.loc[entry].TTTT = True\n",
        "\telse:\n",
        "\t\tguide_candidate_list.loc[entry].TTTT = False\n",
        "\n",
        "\t#Calculate the GC content; we assume higher GC content leads to better annealing and higher Cas9 cutting fidelity.\n",
        "\tguide_candidate_list.loc[entry].GC = (guide_seq.count('C') + guide_seq.count('G')) / len(guide_seq)\n",
        "\n",
        "\t#Determine if the cut will occur far enough away from important genomic features.\n",
        "\t#First find the putative cut site.\n",
        "\tif guide_strand == '+':\n",
        "\t\tputative_cut_site = guide_start + guide_length - cut_site_shift\n",
        "\telse:\n",
        "\t\tputative_cut_site = guide_start + cut_site_shift\n",
        "\n",
        "\t#Then get the critical window based on our padding tolerance.\n",
        "\tcritical_window = range(putative_cut_site - indel_padding, putative_cut_site + indel_padding)\n",
        "\n",
        "\t#Compare the codes of the nucleotides in the window to those on the \"allowed to cut\" list.\n",
        "\twindow_ids = set(seq_code[critical_window])\n",
        "\tif window_ids.issubset(allowed_codes):\n",
        "\t\tguide_candidate_list.loc[entry].safe_to_cut = True\n",
        "\telse:\n",
        "\t\tguide_candidate_list.loc[entry].safe_to_cut = False\n",
        "\n",
        "#Calculate guide specificity.\n",
        "#Specify which n-mers to search for. for SpCas9 and 20bp guide sites, most guide are specific to 20bp. However, it has been shown that binding to the \"seed sequence\"\n",
        "#of roughly 8-10 bp before the PAM site can be enough to trigger nuclease activity. When this occurs at areas different from the designed guide site,\n",
        "#it is known as an off-target effect. We want to choose a guide with the fewest off-targets possible.\n",
        "nmer_list = [8, 10, 12, guide_length]\n",
        "nmer_list.sort(reverse=True)\n",
        "\n",
        "#Add the n-mer counts to the guide dataframe.\n",
        "for num in nmer_list:\n",
        "\tguide_candidate_list[str(num)+'-mer_matches'] = ''\n",
        "\n",
        "#Open the all_guides file that was generated by aphid_prep\n",
        "all_guides_fn = 'all_guides_' + pam_seq + '_' + str(guide_length) + '.txt'\n",
        "all_guides_file = open(all_guides_fn, 'r')\n",
        "all_guides = all_guides_file.read()\n",
        "\n",
        "#Go through each guide candidate we have so far\n",
        "for entry in guide_candidate_list.index:\n",
        "\t#We just need the sequence of the guide\n",
        "\tcandidate_seq = guide_candidate_list.loc[entry].sequence\n",
        "\t#For each n-mer we wanted, find the number of times each truncated guide matches to a PAM site in the genome.\n",
        "\t#The fewer matches, the better.\n",
        "\tfor num in nmer_list:\n",
        "\t\t#Since all the guides are aligned so that the PAM is at the end of the word, we are looking for sequences that match the ends of words.\n",
        "\t\tregex_token = candidate_seq[(-1*num):] + r'\\b'\n",
        "\t\tfield_name = str(num)+'-mer_matches'\n",
        "\t\t#Search for the number of matches.\n",
        "\t\tmatches = len(re.findall(regex_token, all_guides))\n",
        "\t\t#Add the matches to the candidate guide list.\n",
        "\t\tguide_candidate_list.loc[entry, field_name] = matches\n",
        "print('All eligible guides have been annotated.')\n",
        "\n",
        "#We want to sort the guide list so that the \"best\" guide is on top.\n",
        "#To do this, we prioritize guide sequences without poly(T) regions and sequences that are safe to cut.\n",
        "#Then we try and find the most specific guide sequence, and if there is still a tie, we choose the guide with the highest GC content.\n",
        "sort_list = []\n",
        "sort_order_list = []\n",
        "\n",
        "sort_list.append('TTTT')\n",
        "sort_order_list.append(True)\n",
        "\n",
        "sort_list.append('safe_to_cut')\n",
        "sort_order_list.append(False)\n",
        "\n",
        "for num in nmer_list:\n",
        "\tsort_list.append(str(num) + '-mer_matches')\n",
        "\tsort_order_list.append(True)\n",
        "\n",
        "sort_list.append('GC')\n",
        "sort_order_list.append(False)\n",
        "\n",
        "sorted_guide_candidate_list = guide_candidate_list.sort_values(sort_list, ascending = sort_order_list)\n",
        "\n",
        "print('Candidate guides, sorted by quality:')\n",
        "print(sorted_guide_candidate_list.to_string(index=False))\n",
        "print('')\n",
        "\n",
        "#The guide at the top of the list is the one we want to use.\n",
        "best_guide = sorted_guide_candidate_list.iloc[0]\n",
        "\n",
        "print('The best guide:')\n",
        "print(best_guide.to_string())\n",
        "\n",
        "#Provide some warnings in case the best guide is still not ideal for use.\n",
        "if best_guide.TTTT == True:\n",
        "\tprint('ALERT! Best guide is not ideal due to poly(T) presence in sequence. Consider increasing homology arm range to find more suitable guides.')\n",
        "if best_guide.safe_to_cut == False:\n",
        "\tprint('ALERT! Best guide is not ideal due to likely indels in prohibited sequence. Consider increasing homology arm range to find more suitable guides.')\n",
        "if best_guide.GC < 0.3:\n",
        "\tprint('ALERT! Best guide is not ideal due to low GC content. This may not really be an issue, but you could consider increasing homology arm range to find more suitable guides.')\n",
        "field_name = str(nmer_list[1])+'-mer_matches'\n",
        "if best_guide.loc[field_name] > 1:\n",
        "\tprint('ALERT! Best guide is not ideal due to low genome specificty. To avoid off-target effects, consider increasing homology arm range to find more suitable guides.')\n",
        "print('')\n",
        "\n",
        "#DESIGN CLONING PRIMERS\n",
        "\n",
        "#Now that we have chosen the CRISPR guide, we need to design the cloning primers to create the donor template.\n",
        "#The donor template has three pieces, left arm, cargo, and right arm. Each is amplified by a separate pair of primers:\n",
        "#LEFT ARM: L_F_primer_seq + L_R_primer_seq\n",
        "#CARGO: cargo_F_primer_seq + cargo_R_primer_seq\n",
        "#RIGHT ARM: R_F_primer_seq + R_R_primer_seq\n",
        "\n",
        "#We will start with the simplest primers. L_R and R_F flank the insertion site, so there is little variability in what their sequence can be.\n",
        "#We just extend the sequence out from the insertion site until we reach the minimum Tm.\n",
        "#The minimum Tm is defined in aphid_utils but is 59C by default. Unless you are doing something very different or have issues, this does not need to change.\n",
        "\n",
        "print('Designing cloning primers...')\n",
        "#First get the bit for the L_R primer. Note: This is just the genomic portion\n",
        "L_R_genomic = grow_primer_to_tm(local_seq, local_center, '-') #Remember we will RC this later\n",
        "\n",
        "#Now get the bit for the R-F primer. Note: This is just the genomic portion.\n",
        "R_F_genomic = grow_primer_to_tm(local_seq, local_center+1, '+')\n",
        "\n",
        "#The cargo file should only contain a single sequence. Otherwise the program will use the last sequence in the file.\n",
        "#Use the SeqIO docs, there is a function for just getting a single sequence or getting both.\n",
        "for record in SeqIO.parse(cargo_file, \"fasta\"):\n",
        "\t\tcargo_seq = record.seq\n",
        "\n",
        "#Get the GA overlap sequences so that the pieces of the donor template can be assembled with Gibson Assembly or overlapping PCR.\n",
        "L_R_cargo = cargo_seq[0:gibson_assembly_overhang]\n",
        "R_F_cargo = cargo_seq[-1*gibson_assembly_overhang:]\n",
        "\n",
        "#Construct the final primer sequences for R_F and L_R\n",
        "L_R_primer_seq = str(L_R_cargo.reverse_complement()) + str(Seq(L_R_genomic).reverse_complement())\n",
        "R_F_primer_seq = str(R_F_cargo) + str(R_F_genomic)\n",
        "\n",
        "#Now have completed the easiest primers, L-R and R-F. These are constrained by the insertion site and the cargo.\n",
        "\n",
        "#In case the user does not have primers to amplify their cargo sequence, we can design those primers.\n",
        "#First get the forward facing primer to amplify the cargo.\n",
        "cargo_F_primer_seq = grow_primer_to_tm(cargo_seq, 0, '+')\n",
        "\n",
        "#Now get the reverse primer for amplifying cargo.\n",
        "cargo_R_primer_seq = grow_primer_to_tm(cargo_seq, len(cargo_seq), '-').reverse_complement()\n",
        "\n",
        "\n",
        "#Next we want to make the primers for the distal ends of the homology arms, L_F and R_R.\n",
        "#One of these primers will contain the gRNA sequence and PAM, and the other will just be a primer within the homology arm length range.\n",
        "#Get information about our best guide so that we know which arm it is on and where it is located in the local sequence.\n",
        "guide_pos = best_guide.local_position\n",
        "guide_strand = best_guide.strand\n",
        "guide_arm = best_guide.arm\n",
        "\n",
        "#If the guide is on the left arm, then the guide decides the sequence of L_F.\n",
        "if guide_arm == 'left':\n",
        "\t#Start at the end of the guide and grow a primer out to the minimum Tm.\n",
        "\tif guide_strand == '+':\n",
        "\t\tprimer = grow_primer_to_tm(local_seq, guide_pos, '+')\n",
        "\telif guide_strand == '-':\n",
        "\t\tprimer = grow_primer_to_tm(local_seq, guide_pos-len(pam_seq), '+')\n",
        "\tL_F_genomic = primer\n",
        "\t#For R_R, choose a primer that is thermodynamically ideal and is not specific to the genome, to mitigate non-specific bands.\n",
        "\ttry:\n",
        "\t\tR_R_genomic = choose_primer(local_seq, right_lower, right_upper, '-', R_F_primer_seq).sequence\n",
        "\texcept:\n",
        "\t\tprint('Error! Could not find a genomic sequence for R_R. Try adjusting homology arm lengths or primer requirements.')\n",
        "\t\tsys.exit()\n",
        "\n",
        "#If the guide is on the right arm, then the guide decides the sequence of R_R.\n",
        "elif guide_arm == 'right':\n",
        "\tif guide_strand == '+':\n",
        "\t\tprimer = grow_primer_to_tm(local_seq, guide_pos+len(pam_seq)+guide_length-1, '-')\n",
        "\telif guide_strand == '-':\n",
        "\t\tprimer = grow_primer_to_tm(local_seq, guide_pos+guide_length-1, '-')\n",
        "\tR_R_genomic = Seq(primer).reverse_complement()\n",
        "\ttry:\n",
        "\t\tL_F_genomic = choose_primer(local_seq, left_lower, left_upper, '+', L_R_primer_seq).sequence\n",
        "\texcept:\n",
        "\t\tprint('Error! Could not find a genomic sequence for L_F. Try adjusting homology arm lengths or primer requirements.')\n",
        "\t\tsys.exit()\n",
        "#Note: Both R_R genomic sequences have been reverse complemented at this point.\n",
        "\n",
        "#We have the genomic components of L_F and R_R, but in order to clone them into the vector, we need the Gibson Assembly overhangs matching our plasmid.\n",
        "#For pHEE-mCherry, we insert the donor in the HindIII site. The flanking sequence is always the same; I am using 20bp overhangs here.\n",
        "L_plasmid_flanking = 'TAAAACGACGGCCAGTGCCA'\n",
        "R_plasmid_flanking = Seq('TGTTTACACCACAATATATC').reverse_complement()\n",
        "\n",
        "#Combind the genomic and plasmid sequence to create the full primers.\n",
        "L_F_primer_seq = L_plasmid_flanking + L_F_genomic\n",
        "R_R_primer_seq = R_plasmid_flanking + R_R_genomic\n",
        "\n",
        "#We want to calculate where the cloning primers are located within the local sequence so that we can place our genotyping primers accordingly.\n",
        "L_F_indices = range(str(local_seq).find(str(L_F_genomic)), str(local_seq).find(str(L_F_genomic))+len(L_F_genomic))\n",
        "L_R_indices = range(str(local_seq).find(str(L_R_genomic)), str(local_seq).find(str(L_R_genomic))+len(L_R_genomic))\n",
        "\n",
        "R_F_indices = range(str(local_seq).find(str(R_F_genomic)), str(local_seq).find(str(R_F_genomic))+len(R_F_genomic))\n",
        "R_R_indices = range(str(local_seq).find(str(Seq(R_R_genomic).reverse_complement())), str(local_seq).find(str(Seq(R_R_genomic).reverse_complement()))+len(R_R_genomic))\n",
        "\n",
        "print('Designing genotyping primers...')\n",
        "#The primers for cloning the donor element are complete. Now we need two genotyping primers.\n",
        "#The genotyping primers should be outside the homology arms. The distance from the homology arm ends is determined by gt_space_min and gt_space_max.\n",
        "#Non-specific bands can be an issue with genotyping. These primers may not be ideal and you may need to try a few pairs of different\n",
        "#genomic and cargo-specific primers to get consistent and clear results.\n",
        "if gt1_opposite == '':\n",
        "\tgt1_opposite = str(cargo_R_primer_seq)\n",
        "if gt2_opposite == '':\n",
        "\tgt2_opposite = str(cargo_F_primer_seq)\n",
        "\n",
        "#Design the genotyping primers based on their approved location and chosen genotyping partner.\n",
        "try:\n",
        "\tGT1_primer_seq = choose_primer(local_seq, min(L_F_indices)-gt_space_max, min(L_F_indices)-gt_space_min, '+', gt1_opposite).sequence\n",
        "except:\n",
        "\tprint('Error! Could not find a suitable GT1 primer sequence. Try adjusting the gt window or primer requirements.')\n",
        "\tsys.exit()\n",
        "\n",
        "try:\n",
        "\tGT2_primer_seq = choose_primer(local_seq, max(R_R_indices)+gt_space_min, max(R_R_indices)+gt_space_max, '-', gt2_opposite).sequence\n",
        "except:\n",
        "\tprint('Error! Could not find a suitable GT2 primer sequence. Try adjusting the gt window or primer requirements.')\n",
        "\tsys.exit()\n",
        "\n",
        "#Finally we need two primers for the gRNA.\n",
        "#In this protocol, we anneal two primers and insert them into the plasmid via ligation.\n",
        "#Using the pHEE-mCherry vector, this means we are inserting into the BsaI cut site. Therefore each primer needs specific sticky ends.\n",
        "#For improved efficiency, instead of using 20bp gRNA + scaffold, we use G + 19bp gRNA + scaffold\n",
        "gRNA_F_primer_seq = 'ATTG' + best_guide.sequence[1:]\n",
        "gRNA_R_primer_seq = 'AAAC' + str(Seq(best_guide.sequence[1:]).reverse_complement())\n",
        "\n",
        "print('Primer design complete.')\n",
        "print('')\n",
        "\n",
        "#We want to recalculate where the other primers are located within the local sequence so that in our final output we can highlight or underline the primer sequences in context.\n",
        "GT1_indices = range(str(local_seq).find(GT1_primer_seq), str(local_seq).find(GT1_primer_seq)+len(GT1_primer_seq))\n",
        "GT2_indices = range(str(local_seq).find(str(Seq(GT2_primer_seq).reverse_complement())), str(local_seq).find(str(Seq(GT2_primer_seq).reverse_complement()))+len(GT2_primer_seq))\n",
        "\n",
        "#this is probably wrong.\n",
        "gRNA_indices = range(best_guide.local_position, best_guide.local_position + guide_length)\n",
        "\n",
        "print('Final primer sequences:')\n",
        "\n",
        "print(project_name + '_L_F: ' + L_F_primer_seq)\n",
        "print(project_name + '_L_R: ' + L_R_primer_seq)\n",
        "print(project_name + '_R_F: ' + R_F_primer_seq)\n",
        "print(project_name + '_R_R: ' + R_R_primer_seq)\n",
        "\n",
        "print(cargo_name + '_F: ' + cargo_F_primer_seq)\n",
        "print(cargo_name + '_R: ' + cargo_R_primer_seq)\n",
        "\n",
        "print(project_name + '_gRNA_F: ' + gRNA_F_primer_seq)\n",
        "print(project_name + '_gRNA_R: ' + gRNA_R_primer_seq)\n",
        "\n",
        "print(project_name + '_GT1: ' + GT1_primer_seq)\n",
        "print(project_name + '_GT2: ' + GT2_primer_seq)\n",
        "print('')\n",
        "\n",
        "print('Left homology arm length: ' + str(max(L_R_indices)-min(L_F_indices)+1))\n",
        "print('Right homology arm length: ' + str(max(R_R_indices)-min(R_F_indices)+1))\n",
        "print('Cargo sequence length: ' + str(len(cargo_seq)))\n",
        "print('')\n",
        "\n",
        "\n",
        "#OUTPUT FINAL RESULTS\n",
        "\n",
        "print('Outputting final resutls...')\n",
        "#Make a Word Doc with annotated sequence and other information.\n",
        "design_doc = Document()\n",
        "styles = design_doc.styles\n",
        "\n",
        "paragraph = design_doc.add_paragraph()\n",
        "run = paragraph.add_run('APHID Online run with: Gene=' + transcript_name + ' Insertion location=' + str(insertion_location) + ' Cargo File=' + cargo_fn + ' Project Name=' + project_name + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run('Transcript: ' + transcript_name + '\\n' + 'Cargo: ' + cargo_name + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "if insertion_location < 0:\n",
        "\trun = paragraph.add_run('Insertion location: End of coding sequence' + '\\n')\n",
        "else:\n",
        "\trun = paragraph.add_run('Insertion location: After amino acid ' + str(insertion_location) + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run('\\n')\n",
        "\n",
        "\n",
        "for idx,nucleotide in enumerate(str(local_seq)):\n",
        "\tif idx == (local_center+1):\n",
        "\t\tfor nucleotide2 in str(cargo_seq):\n",
        "\t\t\trun = paragraph.add_run(nucleotide2)\n",
        "\t\t\trun.font.color.rgb = RGBColor(0, 255, 0)\n",
        "\t\t\trun.font.name = 'Courier New'\n",
        "\trun = paragraph.add_run(nucleotide)\n",
        "\tmatch seq_code[idx]:\n",
        "\t\tcase 1:\n",
        "\t\t\trun.font.color.rgb = RGBColor(0, 0, 0)\n",
        "\t\tcase 2:\n",
        "\t\t\trun.font.color.rgb = RGBColor(255, 165, 0)\n",
        "\t\tcase 3:\n",
        "\t\t\trun.font.color.rgb = RGBColor(255, 0, 0)\n",
        "\t\tcase 4:\n",
        "\t\t\trun.font.color.rgb = RGBColor(219, 123, 255)\n",
        "\t\tcase 5:\n",
        "\t\t\trun.font.color.rgb = RGBColor(255, 0, 0)\n",
        "\t\tcase 6:\n",
        "\t\t\trun.font.color.rgb = RGBColor(255, 165, 0)\n",
        "\t\tcase _:\n",
        "\t\t\tpass\n",
        "\n",
        "\tif (idx in GT1_indices) or (idx in GT2_indices):\n",
        "\t\trun.font.highlight_color = 16 #Gray\n",
        "\tif (idx in L_F_indices) or (idx in L_R_indices):\n",
        "\t\trun.font.highlight_color = 7 #Yellow\n",
        "\tif (idx in R_F_indices) or (idx in R_R_indices):\n",
        "\t\trun.font.highlight_color = 3 #Turquoise\n",
        "\tif idx in gRNA_indices:\n",
        "\t\trun.font.underline = True\n",
        "\n",
        "\n",
        "\trun.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run('\\n')\n",
        "run = paragraph.add_run('\\n')\n",
        "run = paragraph.add_run('Cloning primers:\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run(project_name + '_L_F: ' + L_F_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run(project_name + '_L_R: ' + L_R_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run(project_name + '_R_F: ' + R_F_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run(project_name + '_R_R: ' + R_R_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run(cargo_name + '_F: ' + cargo_F_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run(cargo_name + '_R: ' + cargo_R_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run(project_name + '_gRNA_F: ' + gRNA_F_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run(project_name + '_gRNA_R: ' + gRNA_R_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run('\\n')\n",
        "run = paragraph.add_run('Genotyping primers:\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run(project_name + '_GT1: ' + GT1_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run(project_name + '_GT2: ' + GT2_primer_seq + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "run = paragraph.add_run('\\n')\n",
        "run = paragraph.add_run('Left homology arm length: ' + str(max(L_R_indices)-min(L_F_indices)+1) + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run('Right homology arm length: ' + str(max(R_R_indices)-min(R_F_indices)+1) + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "run = paragraph.add_run('Cargo sequence length: ' + str(len(cargo_seq)) + '\\n')\n",
        "run.font.name = 'Courier New'\n",
        "\n",
        "paragraph.style = design_doc.styles['No Spacing']\n",
        "design_doc.save(project_name + '_design.docx')\n",
        "\n",
        "#Save the primers to a CSV file which you can open in Excel.\n",
        "primer_data = [[project_name + '_L_F', L_F_primer_seq], [project_name + '_L_R', L_R_primer_seq],\n",
        "\t\t\t\t[project_name + '_R_F', R_F_primer_seq], [project_name + '_R_R', R_R_primer_seq],\n",
        "\t\t\t\t[cargo_name + '_F', cargo_F_primer_seq], [cargo_name + '_R', cargo_R_primer_seq],\n",
        "\t\t\t\t[project_name + '_gRNA_F', gRNA_F_primer_seq], [project_name + '_gRNA_R', gRNA_R_primer_seq],\n",
        "\t\t\t\t[project_name + '_GT1', GT1_primer_seq], [project_name + '_GT1', GT2_primer_seq]]\n",
        "primer_df = pd.DataFrame(primer_data, columns = ['Name', 'Sequence'])\n",
        "primer_df.to_csv(project_name + '_primers.csv', index=False)\n",
        "\n",
        "#Make a FASTA file with the complete plasmid sequence.\n",
        "#Note: This only works if you are using pHEE-mCherry.\n",
        "whole_plasmid_seq = 'Gtttacccgccaatatatcctgtcaaacactgatagtttaaactgaaggcgggaaacgacaatctgatccaagctcaagctgctctagcattcgccattcaggctgcgcaactgttgggaagggcgatcggtgcgggcctcttcgctattacgccagctggcgaaagggggatgtgctgcaaggcgattaagttgggtaacgccagggttttcccagtcacgacgttgtaaaacgacggccagtgcca' \\\n",
        "\t\t\t\t\t+ str(local_seq[min(L_F_indices):(max(L_R_indices)+1)]) \\\n",
        "\t\t\t\t\t+ str(cargo_seq) \\\n",
        "\t\t\t\t\t+ str(local_seq[min(R_F_indices):(max(R_R_indices)+1)]) \\\n",
        "\t\t\t\t\t+ 'tgtttacaccacaatatatcctgccaAGCTGTACGGCTACATCGTGCCGAAGGATGCCCAGATCCTCGTGAACCTCTGGGCCATTGGCAGGGACCCAAACGCCTGGCAGAACGCCGATATTTTCAGCCCAGAGCGCTTCATCGGCTGCGAGATCGATGTTAAGGGCCGCGATTTCGGCCTCCTTCCATTTGGCGCTGGCCGCAGAATTTGCCCAGGCATGAATCTCGCCATCAGGATGCTCACCCTCATGCTCGCCACACTCCTCCAGTTCTTCAACTGGAAGCTCGAAGGCGACATCTCCCCGAAGGACCTCGACATGGACGgtttacccgccaatatatcctgtcatcgacttgccttccgcacaatacatcatttcttcttagctttttttcttcttcttcgttcatacagtttttttttgtttatcagcttacattttcttgaaccgtagctttcgttttcttctttttaactttccattcggagtttttgtatcttgtttcatagtttgtcccaggattagaatgattaggcatcgaaccttcaagaatttgattgaataaaacatcttcattcttaagatatgaagataatcttcaaaaggcccctgggaatctgaaagaagagaagcaggcccatttatatgggaaagaacaatagtatttcttatataggcccatttaagttgaaaacaatcttcaaaagtcccacatcgcttagataagaaaacgaagctgagtttatatacagctagagtcgaagtagtgattg' \\\n",
        "\t\t\t\t\t+ best_guide.sequence \\\n",
        "\t\t\t\t\t+ 'gttttagagctagaaatagcaagttaaaataaggctagtccgttatcaacttgaaaaagtggcaccgagtcggtgcttttttttgcaaaattttccagatcgatttcttcttcctctgttcttcggcgttcaatttctggggttttctcttcgttttctgtaactgaaacctaaaatttgacctaaaaaaaatctcaaataatatgattcagtggttttgtacttttcagttagttgagttttgcagttccgatgagataaaccaataccatggttatactagtgaataaaagcatttgcgtttggtttatcattgcgtttatacaaggacagagatccactgagctggaatagcttaaaaccattatcagaacaaaataaaccattttttgttaagaatcagagcatagtaaacaacagaaacaacctaagagaggtaacttgtccaagaagatagctaattatatctattttataaaagttatcatagtttgtaagtcacaaaagatgcaaataacagagaaactaggagacttgagaatatacattcttgtatatttgtattcgagattgtgaaaatttgaccataagtttaaattcttaaaaagatatatctgatctaggtgatggttatagactgtaattttaccacatgtttaatgatggatagtgacacacatgacacatcgacaacactatagcatcttatttagattacaacatgaaatttttctgtaatacatgtctttgtacataatttaaaagtaattcctaagaaatatatttatacaaggagtttaaagaaaacatagcataaagttcaatgagtagtaaaaaccatatacagtatatagcataaagttcaatgagtttattacaaaagcattggttcactttctgtaacacgacgttaaaccttcgtctccaataggagcgctactgattcaacatgccaatatatactaaatacgtttctacagtcaaatgctttaacgtttcatg' \\\n",
        "\t\t\t\t\t+ 'attaagtgactatttaccgtcaatcctttcccattcctcccactaatccaactttttaattactcttaaatcaccactaagctagtaacgcctatcatgaattagctctactaaatctagcaacctttcaaatttgcagtattgcaggtgtctctgtgtctttaaaatagttgccttatgatttcttcggtttcaagatgatcaaatagttatagatttcatgctcacacatgctcattagatgtgtacatactttacttacccaaatctattttctcgcaaagattttgatggtaaagctgatttggttctattgaactaaatcaaacgagtttcagactgagtgattctaatccggcccattagcccctaaacagacccactaattacgcagcttttaatagagtaattacacctagtttacccactaaaccactaagcactaattatctcacaatctaatgagcttccctcgtaattacttgggctttcactctaccatttatttgtaacagtcaagtctctactgtctctatataaactctctaaagttaacacacaattctcatcacaaacaaatcaaccaaagcaacttctactctttcttctttcgaccttatcaatctgttgagaaatctagatggattacaaggaccacgacggggattacaaggaccacgacattgattacaaggatgatgatgacaagatggctccgaagaagaagaggaaggttggcatccacggggtgccagctgctgacaagaagtactcgatcggcctcgatattgggactaactctgttggctgggccgtgatcaccgacgagtacaaggtgccctcaaagaagttcaaggtcctgggcaacaccgatcggcattccatcaagaagaatctcattggcgctctcctgttcgacagcggcgagacggctgaggctacgcggctcaagcgcaccgcccgcaggcggtacacgcgcaggaagaatcgcatctgctacctgcaggagattttctccaacgag' \\\n",
        "\t\t\t\t\t+ 'atggcgaaggttgacgattctttcttccacaggctggaggagtcattcctcgtggaggaggataagaagcacgagcggcatccaatcttcggcaacattgtcgacgaggttgcctaccacgagaagtaccctacgatctaccatctgcggaagaagctcgtggactccacagataaggcggacctccgcctgatctacctcgctctggcccacatgattaagttcaggggccatttcctgatcgagggggatctcaacccggacaatagcgatgttgacaagctgttcatccagctcgtgcagacgtacaaccagctcttcgaggagaaccccattaatgcgtcaggcgtcgacgcgaaggctatcctgtccgctaggctctcgaagtctcggcgcctcgagaacctgatcgcccagctgccgggcgagaagaagaacggcctgttcgggaatctcattgcgctcagcctggggctcacgcccaacttcaagtcgaatttcgatctcgctgaggacgccaagctgcagctctccaaggacacatacgacgatgacctggataacctcctggcccagatcggcgatcagtacgcggacctgttcctcgctgccaagaatctgtcggacgccatcctcctgtctgatattctcagggtgaacaccgagattacgaaggctccgctctcagcctccatgatcaagcgctacgacgagcaccatcaggatctgaccctcctgaaggcgctggtcaggcagcagctccccgagaagtacaaggagatcttcttcgatcagtcgaagaacggctacgctgggtacattgacggcggggcctctcaggaggagttctacaagttcatcaagccgattctggagaagatggacggcacggaggagctgctggtgaagctcaatcgcgaggacctcctgaggaagcagcggacattcgataacggcagcatcccacaccagattcatctcggggagctgcacgctatcctgaggaggcaggaggacttctacc' \\\n",
        "\t\t\t\t\t+ 'ctttcctcaaggataaccgcgagaagatcgagaagattctgactttcaggatcccgtactacgtcggcccactcgctaggggcaactcccgcttcgcttggatgacccgcaagtcagaggagacgatcacgccgtggaacttcgaggaggtggtcgacaagggcgctagcgctcagtcgttcatcgagaggatgacgaatttcgacaagaacctgccaaatgagaaggtgctccctaagcactcgctcctgtacgagtacttcacagtctacaacgagctgactaaggtgaagtatgtgaccgagggcatgaggaagccggctttcctgtctggggagcagaagaaggccatcgtggacctcctgttcaagaccaaccggaaggtcacggttaagcagctcaaggaggactacttcaagaagattgagtgcttcgattcggtcgagatctctggcgttgaggaccgcttcaacgcctccctggggacctaccacgatctcctgaagatcattaaggataaggacttcctggacaacgaggagaatgaggatatcctcgaggacattgtgctgacactcactctgttcgaggaccgggagatgatcgaggagcgcctgaagacttacgcccatctcttcgatgacaaggtcatgaagcagctcaagaggaggaggtacaccggctgggggaggctgagcaggaagctcatcaacggcattcgggacaagcagtccgggaagacgatcctcgacttcctgaagagcgatggcttcgcgaaccgcaatttcatgcagctgattcacgatgacagcctcacattcaaggaggatatccagaaggctcaggtgagcggccagggggactcgctgcacgagcatatcgcgaacctcgctggctcgccagctatcaagaaggggattctgcagaccgtgaaggttgtggacgagctggtgaaggtcatgggcaggcacaagcctgagaacatcgtcattgagatggcccgggagaatcagaccacgcagaa' \\\n",
        "\t\t\t\t\t+ 'gggccagaagaactcacgcgagaggatgaagaggatcgaggagggcattaaggagctggggtcccagatcctcaaggagcacccggtggagaacacgcagctgcagaatgagaagctctacctgtactacctccagaatggccgcgatatgtatgtggaccaggagctggatattaacaggctcagcgattacgacgtcgatcatatcgttccacagtcattcctgaaggatgactccattgacaacaaggtcctcaccaggtcggacaagaaccggggcaagtctgataatgttccttcagaggaggtcgttaagaagatgaagaactactggcgccagctcctgaatgccaagctgatcacgcagcggaagttcgataacctcacaaaggctgagaggggcgggctctctgagctggacaaggcgggcttcatcaagaggcagctggtcgagacacggcagatcactaagcacgttgcgcagattctcgactcacggatgaacactaagtacgatgagaatgacaagctgatccgcgaggtgaaggtcatcaccctgaagtcaaagctcgtctccgacttcaggaaggatttccagttctacaaggttcgggagatcaacaattaccaccatgcccatgacgcgtacctgaacgcggtggtcggcacagctctgatcaagaagtacccaaagctcgagagcgagttcgtgtacggggactacaaggtttacgatgtgaggaagatgatcgccaagtcggagcaggagattggcaaggctaccgccaagtacttcttctactctaacattatgaatttcttcaagacagagatcactctggccaatggcgagatccggaagcgccccctcatcgagacgaacggcgagacgggggagatcgtgtgggacaagggcagggatttcgcgaccgtcaggaaggttctctccatgccacaagtgaatatcgtcaagaagacagaggtccagactggcgggttctctaaggagtcaattctgcctaag' \\\n",
        "\t\t\t\t\t+ 'cggaacagcgacaagctcatcgcccgcaagaaggactgggatccgaagaagtacggcgggttcgacagccccactgtggcctactcggtcctggttgtggcgaaggttgagaagggcaagtccaagaagctcaagagcgtgaaggagctgctggggatcacgattatggagcgctccagcttcgagaagaacccgatcgatttcctggaggcgaagggctacaaggaggtgaagaaggacctgatcattaagctccccaagtactcactcttcgagctggagaacggcaggaagcggatgctggcttccgctggcgagctgcagaaggggaacgagctggctctgccgtccaagtatgtgaacttcctctacctggcctcccactacgagaagctcaagggcagccccgaggacaacgagcagaagcagctgttcgtcgagcagcacaagcattacctcgacgagatcattgagcagatttccgagttctccaagcgcgtgatcctggccgacgcgaatctggataaggtcctctccgcgtacaacaagcaccgcgacaagccaatcagggagcaggctgagaatatcattcatctcttcaccctgacgaacctcggcgcccctgctgctttcaagtacttcgacacaactatcgatcgcaagaggtacacaagcactaaggaggtcctggacgcgaccctcatccaccagtcgattaccggcctctacgagacgcgcatcgacctgtctcagctcgggggcgacaagcggccagcggcgacgaagaaggcggggcaggcgaagaagaagaagtgagctcagagctttcgttcgtatcatcggtttcgacaacgttcgtcaagttcaatgcatcagtttcattgcgcacacaccagaatcctactgagtttgagtattatggcattgggaaaactgtttttcttgtaccatttgttgtgcttgtaatttactgtgttttttattcggttttcgctatcgaactgtgaaatggaaatggatggag' \\\n",
        "\t\t\t\t\t+ 'aagagttaatgaatgatatggtccttttgttcattctcaaattaatattatttgttttttctcttatttgttgtgtgttgaatttgaaattataagagatatgcaaacattttgttttgagtaaaaatgtgtcaaatcgtggcctctaatgaccgaagttaatatgaggagtaaaacacttgtagttgtaccattatgcttattcactaggcaacaaatatattttcagacctagaaaagctgcaaatgttactgaatacaagtatgtcctcttgtgttttagacatttatgaactttcctttatgtaattttccagaatccttgtcagattctaatcattgctttataattatagttatactcatggatttgtagttgagtatgaaaatattttttaatgcattttatgacttgctaagctggcacaactatatttccaacatcactagctaccatcaaaagattgacttctcatcttactcgattgaaaccaaattaacatagggtttttatttaaataaaagtttaaccttctttttaaaaaattgttcatagtgtcatgtcagaacaagagctacaaatcacacatagcatgcataagcggagctatgatgagtggtattgttttgttcgtcacttgtcactcttttccaacacataatcccgacaacaacgtaagagcatctctctctctccacacacactcatgcatgcatgcattcttacacgtgattgccatgcaaatctcctttctcacctataaatacaaaccaacccttcactacactcttcactcaaaccaaaacaagaaaacatacacaaatagcaaaacggtaccAACAATGGATAACATGGCCATCATCAAGGAGTTCATGCGCTTCAAGGTGCACATGGAGGGCTCCGTGAACGGCCACGAGTTCGAGATCGAGGGCGAGGGCGAGGGCCGCCCCTACGAGGGCACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGT' \\\n",
        "\t\t\t\t\t+ 'CCCCTCAGTTCATGTACGGCTCCAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACTTGAAGCTGTCCTTCCCCGAGGGCTTCAAGTGGGAGCGCGTGATGAACTTCGAGGACGGCGGCGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGCGAGTTCATCTACAAGGTGAAGCTGCGCGGCACCAACTTCCCCTCCGACGGCCCCGTAATGCAGAAGAAGACCATGGGCTGGGAGGCCTCCTCCGAGCGGATGTACCCCGAGGACGGCGCCCTGAAGGGCGAGATCAAGCAGAGGCTGAAGCTGAAGGACGGCGGCCACTACGACGCTGAGGTCAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGCGCCTACAACGTCAACATCAAGTTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAACAGTACGAACGCGCCGAGGGCCGCCACTCCACCGGCGGCATGGACGAGCTGTACAAGTGAgagctcttggactcccatgttggcaaaggcaaccaaacaaacaatgaatgatccgctcctgcatatggggcggtttgagtatttcaactgccatttgggctgaattgaagacatgctcctgtcagaaattccgtgatcttactcaatattcagtaatctcggccaatatcctaaatgtgcgtggctttatctgtctttgtattgtttcatcaattcatgtaacgtttgcttttcttatgaattttcaaataaattatcGtattaattgattgacaacgaattcgtaatcatgtcatagctgtttcctgtgtgaaattgttatccgctcacaattccacacaacatacgagccggaagcataaagtgtaaagcctggggtgcctaatgagtgagctaactcacattaattgcgttgcgctcactgcccgctttccagtcgggaaacctgtcgtgccagctgcattaatgaatcggccaacgcgcggggagaggcggtttgcgtattggctaga' \\\n",
        "\t\t\t\t\t+ 'gcagcttgccaacatggtggagcacgacactctcgtctactccaagaatatcaaagatacagtctcagaagaccaaagggctattgagacttttcaacaaagggtaatatcgggaaacctcctcggattccattgcccagctatctgtcacttcatcaaaaggacagtagaaaaggaaggtggcacctacaaatgccatcattgcgataaaggaaaggctatcgttcaagatgcctctgccgacagtggtcccaaagatggacccccacccacgaggagcatcgtggaaaaagaagacgttccaaccacgtcttcaaagcaagtggattgatgtgataacatggtggagcacgacactctcgtctactccaagaatatcaaagatacagtctcagaagaccaaagggctattgagacttttcaacaaagggtaatatcgggaaacctcctcggattccattgcccagctatctgtcacttcatcaaaaggacagtagaaaaggaaggtggcacctacaaatgccatcattgcgataaaggaaaggctatcgttcaagatgcctctgccgacagtggtcccaaagatggacccccacccacgaggagcatcgtggaaaaagaagacgttccaaccacgtcttcaaagcaagtggattgatgtgatatctccactgacgtaagggatgacgcacaatcccactatccttcgcaagaccttcctctatataaggaagttcatttcatttggagaggacacgctgaaatcaccagtctctctctacaaatctatctctctcgagctttcgcagatcccggggggcaatgagatatgaaaaagcctgaactcaccgcgacgtctgtcgagaagtttctgatcgaaaagttcgacagcgtctccgacctgatgcagctctcggagggcgaagaatctcgtgctttcagcttcgatgtaggagggcgtggatatgtcctgcgggtaaatagctgcgccgatggtttctacaaagatcgttatgtttatcgg' \\\n",
        "\t\t\t\t\t+ 'cactttgcatcggccgcgctcccgattccggaagtgcttgacattggggagtttagcgagagcctgacctattgcatctcccgccgtgcacagggtgtcacgttgcaagacctgcctgaaaccgaactgcccgctgttctacaaccggtcgcggaggctatggatgcgatcgctgcggccgatcttagccagacgagcgggttcggcccattcggaccgcaaggaatcggtcaatacactacatggcgtgatttcatatgcgcgattgctgatccccatgtgtatcactggcaaactgtgatggacgacaccgtcagtgcgtccgtcgcgcaggctctcgatgagctgatgctttgggccgaggactgccccgaagtccggcacctcgtgcacgcggatttcggctccaacaatgtcctgacggacaatggccgcataacagcggtcattgactggagcgaggcgatgttcggggattcccaatacgaggtcgccaacatcttcttctggaggccgtggttggcttgtatggagcagcagacgcgctacttcgagcggaggcatccggagcttgcaggatcgccacgactccgggcgtatatgctccgcattggtcttgaccaactctatcagagcttggttgacggcaatttcgatgatgcagcttgggcgcagggtcgatgcgacgcaatcgtccgatccggagccgggactgtcgggcgtacacaaatcgcccgcagaagcgcggccgtctggaccgatggctgtgtagaagtactcgccgatagtggaaaccgacgccccagcactcgtccgagggcaaagaaatagagtagatgccgaccggatctgtcgatcgacaagctcgagtttctccataataatgtgtgagtagttcccagataagggaattagggttcctatagggtttcgctcatgtgttgagcatataagaaacccttagtatgtatttgtatttgtaaaatacttctatcaataaaatttctaattcctaaaaccaaaa' \\\n",
        "\t\t\t\t\t+ 'tccagtactaaaatccagatcccccgaattaattcggcgttaattcagtacattaaaaacgtccgcaatgtgttattaagttgtctaagcgtcaatttgtttacaccacaatatatcctgccaccagccagccaacagctccccgaccggcagctcggcacaaaatcaccactcgatacaggcagcccatcagtccgggacggcgtcagcgggagagccgttgtaaggcggcagactttgctcatgttaccgatgctattcggaagaacggcaactaagctgccgggtttgaaacacggatgatctcgcggagggtagcatgttgattgtaacgatgacagagcgttgctgcctgtgatcaccgcggtttcaaaatcggctccgtcgatactatgttatacgccaactttgaaaacaactttgaaaaagctgttttctggtatttaaggttttagaatgcaaggaacagtgaattggagttcgtcttgttataattagcttcttggggtatctttaaatactgtagaaaagaggaaggaaataataaatggctaaaatgagaatatcaccggaattgaaaaaactgatcgaaaaataccgctgcgtaaaagatacggaaggaatgtctcctgctaaggtatataagctggtgggagaaaatgaaaacctatatttaaaaatgacggacagccggtataaagggaccacctatgatgtggaacgggaaaaggacatgatgctatggctggaaggaaagctgcctgttccaaaggtcctgcactttgaacggcatgatggctggagcaatctgctcatgagtgaggccgatggcgtcctttgctcggaagagtatgaagatgaacaaagccctgaaaagattatcgagctgtatgcggagtgcatcaggctctttcactccatcgacatatcggattgtccctatacgaatagcttagacagccgcttagccgaattggattacttactgaataacgatctggccgatgtggattgcgaaaactgg' \\\n",
        "\t\t\t\t\t+ 'gaagaagacactccatttaaagatccgcgcgagctgtatgattttttaaagacggaaaagcccgaagaggaacttgtcttttcccacggcgacctgggagacagcaacatctttgtgaaagatggcaaagtaagtggctttattgatcttgggagaagcggcagggcggacaagtggtatgacattgccttctgcgtccggtcgatcagggaggatatcggggaagaacagtatgtcgagctattttttgacttactggggatcaagcctgattgggagaaaataaaatattatattttactggatgaattgttttagtacctagaatgcatgaccaaaatcccttaacgtgagttttcgttccactgagcgtcagaccccgtagaaaagatcaaaggatcttcttgagatcctttttttctgcgcgtaatctgctgcttgcaaacaaaaaaaccaccgctaccagcggtggtttgtttgccggatcaagagctaccaactctttttccgaaggtaactggcttcagcagagcgcagataccaaatactgtccttctagtgtagccgtagttaggccaccacttcaagaactctgtagcaccgcctacatacctcgctctgctaatcctgttaccagtggctgctgccagtggcgataagtcgtgtcttaccgggttggactcaagacgatagttaccggataaggcgcagcggtcgggctgaacggggggttcgtgcacacagcccagcttggagcgaacgacctacaccgaactgagatacctacagcgtgagctatgagaaagcgccacgcttcccgaagggagaaaggcggacaggtatccggtaagcggcagggtcggaacaggagagcgcacgagggagcttccagggggaaacgcctggtatctttatagtcctgtcgggtttcgccacctctgacttgagcgtcgatttttgtgatgctcgtcaggggggcggagcctatggaaaaacgccagcaacgcggcctttttacggttcc' \\\n",
        "\t\t\t\t\t+ 'tggccttttgctggccttttgctcacatgttctttcctgcgttatcccctgattctgtggataaccgtattaccgcctttgagtgagctgataccgctcgccgcagccgaacgaccgagcgcagcgagtcagtgagcgaggaagcggaagagcgcctgatgcggtattttctccttacgcatctgtgcggtatttcacaccgcatatggtgcactctcagtacaatctgctctgatgccgcatagttaagccagtatacactccgctatcgctacgtgactgggtcatggctgcgccccgacacccgccaacacccgctgacgcgccctgacgggcttgtctgctcccggcatccgcttacagacaagctgtgaccgtctccgggagctgcatgtgtcagaggttttcaccgtcatcaccgaaacgcgcgaggcagggtgccttgatgtgggcgccggcggtcgagtggcgacggcgcggcttgtccgcgccctggtagattgcctggccgtaggccagccatttttgagcggccagcggccgcgataggccgacgcgaagcggcggggcgtagggagcgcagcgaccgaagggtaggcgctttttgcagctcttcggctgtgcgctggccagacagttatgcacaggccaggcgggttttaagagttttaataagttttaaagagttttaggcggaaaaatcgccttttttctcttttatatcagtcacttacatgtgtgaccggttcccaatgtacggctttgggttcccaatgtacgggttccggttcccaatgtacggctttgggttcccaatgtacgtgctatccacaggaaacagaccttttcgacctttttcccctgctagggcaatttgccctagcatctgctccgtacattaggaaccggcggatgcttcgccctcgatcaggttgcggtagcgcatgactaggatcgggccagcctgccccgcctcctccttcaaatcgtactccggcaggtcatttgacccgatcagcttgcg' \\\n",
        "\t\t\t\t\t+ 'cacggtgaaacagaacttcttgaactctccggcgctgccactgcgttcgtagatcgtcttgaacaaccatctggcttctgccttgcctgcggcgcggcgtgccaggcggtagagaaaacggccgatgccgggatcgatcaaaaagtaatcggggtgaaccgtcagcacgtccgggttcttgccttctgtgatctcgcggtacatccaatcagctagctcgatctcgatgtactccggccgcccggtttcgctctttacgatcttgtagcggctaatcaaggcttcaccctcggataccgtcaccaggcggccgttcttggccttcttcgtacgctgcatggcaacgtgcgtggtgtttaaccgaatgcaggtttctaccaggtcgtctttctgctttccgccatcggctcgccggcagaacttgagtacgtccgcaacgtgtggacggaacacgcggccgggcttgtctcccttcccttcccggtatcggttcatggattcggttagatgggaaaccgccatcagtaccaggtcgtaatcccacacactggccatgccggccggccctgcggaaacctctacgtgcccgtctggaagctcgtagcggatcacctcgccagctcgtcggtcacgcttcgacagacggaaaacggccacgtccatgatgctgcgactatcgcgggtgcccacgtcatagagcatcggaacgaaaaaatctggttgctcgtcgcccttgggcggcttcctaatcgacggcgcaccggctgccggcggttgccgggattctttgcggattcgatcagcggccgcttgccacgattcaccggggcgtgcttctgcctcgatgcgttgccgctgggcggcctgcgcggccttcaacttctccaccaggtcatcacccagcgccgcgccgatttgtaccgggccggatggtttgcgaccgctcacgccgattcctcgggcttgggggttccagtgccattgcagggccggcagacaacccagccgcttacgcctggccaac' \\\n",
        "\t\t\t\t\t+ 'cgcccgttcctccacacatggggcattccacggcgtcggtgcctggttgttcttgattttccatgccgcctcctttagccgctaaaattcatctactcatttattcatttgctcatttactctggtagctgcgcgatgtattcagatagcagctcggtaatggtcttgccttggcgtaccgcgtacatcttcagcttggtgtgatcctccgccggcaactgaaagttgacccgcttcatggctggcgtgtctgccaggctggccaacgttgcagccttgctgctgcgtgcgctcggacggccggcacttagcgtgtttgtgcttttgctcattttctctttacctcattaactcaaatgagttttgatttaatttcagcggccagcgcctggacctcgcgggcagcgtcgccctcgggttctgattcaagaacggttgtgccggcggcggcagtgcctgggtagctcacgcgctgcgtgatacgggactcaagaatgggcagctcgtacccggccagcgcctcggcaacctcaccgccgatgcgcgtgcctttgatcgcccgcgacacgacaaaggccgcttgtagccttccatccgtgacctcaatgcgctgcttaaccagctccaccaggtcggcggtggcccatatgtcgtaagggcttggctgcaccggaatcagcacgaagtcggctgccttgatcgcggacacagccaagtccgccgcctggggcgctccgtcgatcactacgaagtcgcgccggccgatggccttcacgtcgcggtcaatcgtcgggcggtcgatgccgacaacggttagcggttgatcttcccgcacggccgcccaatcgcgggcactgccctggggatcggaatcgactaacagaacatcggccccggcgagttgcagggcgcgggctagatgggttgcgatggtcgtcttgcctgacccgcctttctggttaagtacagcgataaccttcatgcgttccccttgcgtatttgtttatttactcatcgcatcata' \\\n",
        "\t\t\t\t\t+ 'tacgcagcgaccgcatgacgcaagctgttttactcaaatacacatcacctttttagacggcggcgctcggtttcttcagcggccaagctggccggccaggccgccagcttggcatcagacaaaccggccaggatttcatgcagccgcacggttgagacgtgcgcgggcggctcgaacacgtacccggccgcgatcatctccgcctcgatctcttcggtaatgaaaaacggttcgtcctggccgtcctggtgcggtttcatgcttgttcctcttggcgttcattctcggcggccgccagggcgtcggcctcggtcaatgcgtcctcacggaaggcaccgcgccgcctggcctcggtgggcgtcacttcctcgctgcgctcaagtgcgcggtacagggtcgagcgatgcacgccaagcagtgcagccgcctctttcacggtgcggccttcctggtcgatcagctcgcgggcgtgcgcgatctgtgccggggtgagggtagggcgggggccaaacttcacgcctcgggccttggcggcctcgcgcccgctccgggtgcggtcgatgattagggaacgctcgaactcggcaatgccggcgaacacggtcaacaccatgcggccggccggcgtggtggtgtcggcccacggctctgccaggctacgcaggcccgcgccggcctcctggatgcgctcggcaatgtccagtaggtcgcgggtgctgcgggccaggcggtctagcctggtcactgtcacaacgtcgccagggcgtaggtggtcaagcatcctggccagctccgggcggtcgcgcctggtgccggtgatcttctcggaaaacagcttggtgcagccggccgcgtgcagttcggcccgttggttggtcaagtcctggtcgtcggtgctgacgcgggcatagcccagcaggccagcggcggcgctcttgttcatggcgtaatgtctccggttctagtcgcaagtattctactttatgcgactaaaacacgcgacaagaaaacgccaggaaaaggg' \\\n",
        "\t\t\t\t\t+ 'cagggcggcagcctgtcgcgtaacttaggacttgtgcgacatgtcgttttcagaagacggctgcactgaacgtcagaagccgactgcactatagcagcggaggggttggatcaaagtactttgatcccgaggggaaccctgtggttggcatgcacatacaaatggacgaacggataaaccttttcacgcccttttaaatatccgttattctaataaacgctcttttctcttag'\n",
        "with open(project_name + '_plasmid.fa', 'w') as f:\n",
        "\tf.write('>pHEE_mCherry_' + project_name + '\\n')\n",
        "\tf.write(whole_plasmid_seq)\n",
        "\n",
        "print('Design complete. Your three output files should be in the folder to the left\\n<---')"
      ],
      "metadata": {
        "id": "9azEdztFO6rC",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6544510a-7dbc-49e2-d53c-600592ac5286",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Transcript ID: AT1G73590.1\n",
            "Chromosome number: 1\n",
            "Extracting transcript features...\n",
            "Complete.\n",
            "\n",
            "Desired insertion is after residue number: 452.\n",
            "Insertion location is at coordinate 27661279 on Chromosome 1.\n",
            "\n",
            "Extracting sequence flanking the insertion site...\n",
            "Annotating the sequence around the insertion site...\n",
            "Complete.\n",
            "\n",
            "Searching for eligible guide sequences within the desired homology arm regions...\n",
            "Annotating eligible guides and calculating specificity...\n",
            "All eligible guides have been annotated.\n",
            "Candidate guides, sorted by quality:\n",
            "            sequence local_position strand   arm    GC  TTTT safe_to_cut 20-mer_matches 12-mer_matches 10-mer_matches 8-mer_matches\n",
            "AAAAACACCATAAAGAATGA           1999      - right  0.25 False        True              1              1             23           362\n",
            "ATTATGTAATAAATGAAAAG           1936      - right  0.15 False        True              1              3             63           370\n",
            "TTTCTGACCTTCATTCTTTA           1989      + right   0.3 False        True              1              5             26           314\n",
            "AAGAAAAAAAACAAATCATG           2037      + right   0.2 False        True              1              8             28           374\n",
            "AAATAAAAGAAAAGAGAAAA           1910      - right  0.15 False        True              1             24            188          1291\n",
            "AACATCCCTGGGTTCGGCGC            980      -  left  0.65 False       False              1              1              1            36\n",
            "TAGCGGTGGAGTAATCGGCG           1165      -  left   0.6 False       False              1              1              2            32\n",
            "GCCTCCACCACCGCCAGTGT           1008      -  left   0.7 False       False              1              1              2            50\n",
            "GGCGAAAACATCCCTGGGTT            986      -  left  0.55 False       False              1              1              2            56\n",
            "GTGGAGCTCAAGTGCTTCGC           1105      +  left   0.6 False       False              1              1              2            76\n",
            "CCTCCACCACCGCCAGTGTT           1007      -  left  0.65 False       False              1              1              2            81\n",
            "GTCTCAGATGTGTTCGGCGG           1127      +  left   0.6 False       False              1              1              2            93\n",
            "CGGTGGAGGAGCGCATTATC            958      +  left   0.6 False       False              1              1              2           129\n",
            "CATTATCCGGCGCCGAACCC            971      +  left  0.65 False       False              1              1              3            15\n",
            "TTGATGATCGTTCGTAGCGG           1179      -  left   0.5 False       False              1              1              5            39\n",
            "GGCGGCGAAAGGAAACGCTC           1030      +  left  0.65 False       False              1              1              5            56\n",
            "ATTATCCGGCGCCGAACCCA            972      +  left   0.6 False       False              1              1              5            65\n",
            "ATAGACTTTGCTATAAGAGC           1833      - right  0.35 False       False              1              1              5           106\n",
            "CTCAGATGCAGGTCTAGGCA           1862      + right  0.55 False       False              1              1              6           169\n",
            "AAAGGAAACGCTCCGGTGGT           1037      +  left  0.55 False       False              1              1              8           125\n",
            "CGGTGGAGTAATCGGCGTGG           1162      -  left  0.65 False       False              1              1             10            48\n",
            "CCGCCGAACACATCTGAGAC           1127      -  left   0.6 False       False              1              1             10           117\n",
            "GGTGGAGGCACGGCGGCGAA           1019      +  left  0.75 False       False              1              1             11            58\n",
            "GAGATCTTCACATGTTTGTG           1086      +  left   0.4 False       False              1              1             11           441\n",
            "GAAACGCTCCGGTGGTTGGT           1041      +  left   0.6 False       False              1              1             15           328\n",
            "GGGAAAAGACAAGACGGAAA           1061      +  left  0.45 False       False              1              1             18           127\n",
            "GGGAGGTTTCATTATCAATC            923      +  left   0.4 False       False              1              1             18           169\n",
            "TCGCCGGTCTCAGATGTGTT           1121      +  left  0.55 False       False              1              1             18           359\n",
            "GAACTGCTGCTGGAGCTGGG            906      +  left  0.65 False       False              1              2              4            62\n",
            "GTGTTGGGCGAAAACATCCC            992      -  left  0.55 False       False              1              2              4            67\n",
            "TGGCGGTGGTGGAGGCACGG           1012      +  left  0.75 False       False              1              2              5            41\n",
            "GGCATGGCTATGTTCAGTCT           1878      + right   0.5 False       False              1              2              5           140\n",
            "ATACTCTCAGATGCAGGTCT           1857      + right  0.45 False       False              1              2              8            67\n",
            "CACTGGCGGTGGTGGAGGCA           1009      +  left   0.7 False       False              1              2              9           111\n",
            "GTTCGTAGCGGTGGAGTAAT           1170      -  left   0.5 False       False              1              2             11           142\n",
            "GGAAACGCTCCGGTGGTTGG           1040      +  left  0.65 False       False              1              2             13           203\n",
            "CATTATCAATCTGGAGGAAG            932      +  left   0.4 False       False              1              2             22           742\n",
            "GTTGGTGGGAAAAGACAAGA           1055      +  left  0.45 False       False              1              2             39           339\n",
            "GGCGAAAGGAAACGCTCCGG           1033      +  left  0.65 False       False              1              3              6            40\n",
            "CCGGTCTCAGATGTGTTCGG           1124      +  left   0.6 False       False              1              3              8            44\n",
            "TGTTGGGCGAAAACATCCCT            991      -  left   0.5 False       False              1              3              9            82\n",
            "TAGACCTGCATCTGAGAGTA           1858      - right  0.45 False       False              1              3             11           143\n",
            "CTGGAACTGCTGCTGGAGCT            903      +  left   0.6 False       False              1              3             13           119\n",
            "GCTGGAACTGCTGCTGGAGC            902      +  left  0.65 False       False              1              3             15           165\n",
            "AGGTTTCATTATCAATCTGG            926      +  left  0.35 False       False              1              3             17           120\n",
            "TCAGATGTGTTCGGCGGTGG           1130      +  left   0.6 False       False              1              3             33           798\n",
            "TCTGGAGGAAGTGGTGGCGG            941      +  left  0.65 False       False              1              3            172           580\n",
            "TCGCCCAACACTGGCGGTGG           1001      +  left   0.7 False       False              1              4             27           798\n",
            "TATCAATCTGGAGGAAGTGG            935      +  left  0.45 False       False              1              6             73           397\n",
            "CAATCTGGAGGAAGTGGTGG            938      +  left  0.55 False       False              1             16             59           572\n",
            "GATGTGTTCGGCGGTGGAGG           1133      +  left  0.65 False       False              1             29            134          1142\n",
            "GGAGGAAGTGGTGGCGGTGG            944      +  left   0.7 False       False              1             64            182           798\n",
            "CCCAACACTGGCGGTGGTGG           1004      +  left   0.7 False       False              1             64            244          2915\n",
            "ATCTCCATACTCTCAGATGC           1851      + right  0.45 False       False              2              2              9           125\n",
            "GAGAACAATTGACTGTTTTA           1965      - right   0.3  True        True              1              1              8           270\n",
            "GGGATGTTTTCGCCCAACAC            992      +  left  0.55  True       False              1              1              1            54\n",
            "TTTTCGCCCAACACTGGCGG            998      +  left   0.6  True       False              1              1              2            78\n",
            "ATGTTTTCGCCCAACACTGG            995      +  left   0.5  True       False              1              1              6           107\n",
            "TGTCTTTTCCCACCAACCAC           1052      -  left   0.5  True       False              1              3             12           101\n",
            "\n",
            "The best guide:\n",
            "sequence          AAAAACACCATAAAGAATGA\n",
            "local_position                    1999\n",
            "strand                               -\n",
            "arm                              right\n",
            "GC                                0.25\n",
            "TTTT                             False\n",
            "safe_to_cut                       True\n",
            "20-mer_matches                       1\n",
            "12-mer_matches                       1\n",
            "10-mer_matches                      23\n",
            "8-mer_matches                      362\n",
            "ALERT! Best guide is not ideal due to low GC content. This may not really be an issue, but you could consider increasing homology arm range to find more suitable guides.\n",
            "\n",
            "Designing cloning primers...\n",
            "Designing genotyping primers...\n",
            "Primer design complete.\n",
            "\n",
            "Final primer sequences:\n",
            "PIN1-GFP_L_F: TAAAACGACGGCCAGTGCCAGGTGGTTGGTGGGAAAAGAC\n",
            "PIN1-GFP_L_R: TCCTCGCCCTTGCTCACCATAGCCTGCGTCGTTTTGTTGC\n",
            "PIN1-GFP_R_F: GCATGGACGAGCTGTACAAGAAGGTGATGCCACCAACAAGTG\n",
            "PIN1-GFP_R_R: GATATATTGTGGTGTAAACAAAAAACACCATAAAGAATGAAGGTCAGAAATAAA\n",
            "GFP_F: ATGGTGAGCAAGGGCGAGG\n",
            "GFP_R: CTTGTACAGCTCGTCCATGCC\n",
            "PIN1-GFP_gRNA_F: ATTGAAAACACCATAAAGAATGA\n",
            "PIN1-GFP_gRNA_R: AAACTCATTCTTTATGGTGTTTT\n",
            "PIN1-GFP_GT1: GACCTTCCAACTACGAAGAAGA\n",
            "PIN1-GFP_GT2: TGTGAAGCCAGTTTTGTTTGC\n",
            "\n",
            "Left homology arm length: 450\n",
            "Right homology arm length: 518\n",
            "Cargo sequence length: 717\n",
            "\n",
            "Outputting final resutls...\n",
            "Design complete. Your three output files should be in the folder to the left\n",
            "<---\n"
          ]
        }
      ]
    }
  ]
}